File: adk/go/weil_wallet/api/platform_api.go
package api

import (
	"bytes"
	"compress/gzip"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/textproto"

	"github.com/weilliptic-inc/contract-sdk/go/weil_wallet/internal/constants"
	"github.com/weilliptic-inc/contract-sdk/go/weil_wallet/transaction"
)

func SubmitTransaction(httpClient *http.Client, payload SubmitTxnRequest, isNonBlocking bool) (*transaction.TransactionResult, error) {
	// Compress payload
	payloadJson, err := json.Marshal(payload)

	if err != nil {
		return nil, err
	}

	var buf bytes.Buffer
	w := gzip.NewWriter(&buf)

	if _, err := w.Write(payloadJson); err != nil {
		return nil, fmt.Errorf("payload compression failed: %w", err)
	}

	if err := w.Close(); err != nil {
		return nil, fmt.Errorf("payload compression failed: %w", err)
	}

	// Build multipart form
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)

	part, err := writer.CreatePart(textproto.MIMEHeader{
		"Content-Disposition": []string{`form-data; name="transaction"; filename="transaction_data"`},
		"Content-Type":        []string{"application/octet-stream"},
	})

	if err != nil {
		return nil, fmt.Errorf("failed to create multipart part: %w", err)
	}

	if _, err := part.Write(buf.Bytes()); err != nil {
		return nil, fmt.Errorf("failed to write multipart part: %w", err)
	}

	writer.Close()

	// Build request
	req, err := http.NewRequest(http.MethodPost,
		fmt.Sprintf("https://%s/contracts/execute_smartcontract", constants.SENTINEL_HOST),
		body,
	)

	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", writer.FormDataContentType())

	if isNonBlocking {
		req.Header.Set("x-non-blocking", "true")
	}

	resp, err := httpClient.Do(req)

	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("failed to submit the transaction")
	}

	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)

	if err != nil {
		return nil, err
	}

	var txnResult transaction.TransactionResult

	if isNonBlocking {
		return &transaction.TransactionResult{}, nil
	}

	if json.Unmarshal(respBody, &txnResult) != nil {
		return nil, err
	}

	return &txnResult, nil
}

File: adk/python/weil_wallet/api/platform_api.py
"""Platform API: submit transaction (normal and streaming)."""

from typing import AsyncIterator

import httpx

from ..transaction import TransactionResult
from ..utils import compress
from .request import SubmitTxnRequest


class PlatformApi:
    """Submit transactions to the WeilChain platform."""

    @staticmethod
    async def _submit_transaction_inner(
        payload: SubmitTxnRequest,
        client: httpx.AsyncClient,
        *,
        is_non_blocking: bool,
    ) -> httpx.Response:
        """GZIP-compress and POST the transaction payload; raise on HTTP error."""
        payload_dict = payload.to_payload_dict()
        tx_payload = compress(payload_dict)

        files = {
            "transaction": ("transaction_data", tx_payload, "application/octet-stream")
        }

        headers = {}

        if is_non_blocking:
            headers["x-non-blocking"] = "true"

        response = await client.post(
            "/contracts/execute_smartcontract", files=files, headers=headers
        )

        if not response.is_success:
            body = response.text[:500] if response.text else ""

            raise RuntimeError(
                f"failed to submit the transaction: HTTP {response.status_code} {body}"
            )

        return response

    @staticmethod
    async def submit_transaction(
        payload: SubmitTxnRequest,
        client: httpx.AsyncClient,
        *,
        is_non_blocking: bool = False,
    ) -> TransactionResult:
        """Submit and return the parsed TransactionResult."""
        response = await PlatformApi._submit_transaction_inner(
            payload, client, is_non_blocking=is_non_blocking
        )
        data = response.json()
        return TransactionResult.from_dict(data)

    @staticmethod
    async def submit_transaction_with_streaming(
        payload: SubmitTxnRequest,
        client: httpx.AsyncClient,
        *,
        is_non_blocking: bool = False,
    ) -> AsyncIterator[bytes]:
        """Submit and return an async iterator of response body chunks."""
        response = await PlatformApi._submit_transaction_inner(
            payload, client, is_non_blocking=is_non_blocking
        )
        async for chunk in response.aiter_bytes():
            if chunk:
                yield chunk
